File: 78. Subsets.py
class Solution:
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        if len(nums) == 0:
            return [[]]
        if len(nums) == 1:
            return [nums, []]
        l1 = self.subsets(nums[1:])
        l2 = l1.copy()
        for i in range(len(l2)):
            l2[i] = [nums[0]] + l2[i]
        return l2 + l1


sol = Solution()
print(sol.subsets([1, 2, 3, 4]))
print(sol.subsets([1, 2, 3]))
print(sol.subsets([1, 2]))
print(sol.subsets([1]))
print(sol.subsets([]))

File: 78. Subsets.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> curr;
        for (size_t i=0; i <= nums.size(); ++i) {
            backtracking(nums, i, 0, curr, ans);
        }
        return ans;
    }

    void backtracking(vector<int>& nums, size_t subset_len, size_t level, vector<int>& curr, vector<vector<int>>& ans) {
        if (curr.size() == subset_len) {
            ans.push_back(curr);
            return;
        }
        for (size_t i=level; i < nums.size(); ++i) {
            curr.push_back(nums[i]);
            backtracking(nums, subset_len, i+1, curr, ans);
            curr.pop_back();
        }
    }
};

int main() {
    Solution a;
    vector<int> v{1,2,3};
    vector<vector<int>> ans = a.subsets(v);
    cout << "Size: " << ans.size() << endl;
    for (auto &vec: ans) {
        for (auto element: vec) {
            cout << element << ' ';
        }
        cout << endl;
    }
    return 0;
}
